File: src/code-references/IMSortAlgorithms/src/com/ivanmagda/sort/MergeSort.java
package com.ivanmagda.sort;

public class MergeSort {

    private MergeSort() {
    }

    public static void sort(int[] array) {
        sort(array, 0, array.length - 1);
    }

    private static void sort(int[] array, int start, int end) {
        if (start < end) {
            int middle = start + (end - start) / 2;

            sort(array, start, middle);
            sort(array, middle + 1, end);
            merge(array, start, middle, end);
        }
    }

    private static void merge(int[] array, int start, int middle, int end) {
        int countLeft = middle - start + 1;
        int countRight = end - middle;

        int[] leftArr = new int[countLeft];
        int[] rightArr = new int[countRight];

        System.arraycopy(array, start, leftArr, 0, countLeft);
        System.arraycopy(array, middle + 1, rightArr, 0, countRight);

        int lhsIdx = 0, rhsIdx = 0, copyIndex = start;
        while (lhsIdx < countLeft && rhsIdx < countRight) {
            if (leftArr[lhsIdx] <= rightArr[rhsIdx]) {
                array[copyIndex++] = leftArr[lhsIdx++];
            } else {
                array[copyIndex++] = rightArr[rhsIdx++];
            }
        }

        System.arraycopy(leftArr, lhsIdx, array, copyIndex, countLeft - lhsIdx);
        System.arraycopy(rightArr, rhsIdx, array, copyIndex, countRight - rhsIdx);
    }
}

File: src/code-references/ExternalMergeSort/src/sort/MergeSort.kt
package sort

import model.Sortable

class MergeSort : Sortable<Int> {
    override fun sort(array: Array<Int>) {
        sort(array, Array(array.size, { 0 }), 0, array.lastIndex)
    }

    private fun sort(array: Array<Int>, buffer: Array<Int>, start: Int, end: Int) {
        if (start < end) {
            val middleIndex = start + (end - start) / 2

            sort(array, buffer, start, middleIndex)
            sort(array, buffer, middleIndex + 1, end)
            merge(array, buffer, start, middleIndex, end)
        }
    }

    private fun merge(array: Array<Int>, buffer: Array<Int>, start: Int, middle: Int, end: Int) {
        var copyIndex = start
        var left = start
        var right = middle + 1

        while (left <= middle && right <= end) {
            if (array[left] <= array[right]) {
                buffer[copyIndex++] = array[left++]
            } else {
                buffer[copyIndex++] = array[right++]
            }
        }

        System.arraycopy(array, left, buffer, copyIndex, middle - left + 1)
        System.arraycopy(array, right, buffer, copyIndex, end - right + 1)
        System.arraycopy(buffer, start, array, start, end - start + 1)
    }
}
